cmsTriton bug fixes and improvements

kpedro88 opened the pull request:
#### PR description:

Rollup of several pending bug fixes and improvements, primarily aimed at the `cmsTriton` server management script:
1. Some relval jobs have been seeing timeouts when starting the local fallback server if the server image (hosted on `/cvms/unpacked.cern.ch`) is not already loaded in the worker node's cache. It does not appear to be possible for a non-privileged user to check if something is in the cvmfs cache or not, so the default server start timeout is just extended to minimize these failures.
2. Thread control settings for the local fallback server are only used in the CPU case, not the GPU case.
3. The OSG apptainer version is used by default if available. This avoids the need to propagate updated apptainer versions through a slew of different containers.
4. The SONIC special workflows are extended to 14 TeV processes for Run 3.
5. The `TritonService` will print the local fallback server log if the process is terminated because of an exception that could be related to the local fallback server. (This is denoted by a third argument in the `TRITON_THROW_IF_ERROR` macro.) During inference requests, which can be retried, an exception does not necessarily lead to process termination, so the notification to the `TritonService` can be undone in that case.
6. The unit of the inference request timeout can now be selected. (This was useful for testing the above feature.)
7. More granular verbosity control in the unit test script is now available; clients, the fallback server, and the `TritonService` itself can all be targeted separately.
8. The unit test script is migrated from `VarParsing` to `argparse`, now that the latter works properly with `cmsRun`.
9. The GPU driver check is skipped if the local fallback server is not using GPU.
10. Fix handling of symbolic links in `checksum` and `versioncheck` modes of `cmsTritonConfigTool`
11. Avoid throwing exceptions in `TritonMemResource` destructors (addresses #38260)

#### PR validation:

Unit tests pass.

The following command was used to test item 5:
```
cmsRun test/tritonTest_cfg.py --modules TritonGraphProducer --maxEvents 10 --timeout 1 --timeoutUnit microseconds --verboseService --device CPU --noShm
```

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Not intended to be backported.

kpedro88 committed:
extend default server startup timeout to account for non-cached case

kpedro88 committed:
do not apply thread control settings if GPU use is requested

kpedro88 committed:
use OSG apptainer by default (if available)

kpedro88 committed:
cleanup auto_stop

kpedro88 committed:
extend sonic Run3 workflows to 14 TeV

kpedro88 committed:
fix typo

kpedro88 committed:
notify TritonService when a TritonException is issued (in order to save logs)

kpedro88 committed:
move signal undo

kpedro88 committed:
print log instead of copying

kpedro88 committed:
only notify when error is related to local server

kpedro88 committed:
select timeout unit

kpedro88 committed:
more granular verbosity control

kpedro88 committed:
argparse migration for tritonTest_cfg

kpedro88 committed:
skip driver check if not using GPU

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"0c315ee5961281bb813f51450c3d03e653efc4d9":{"files":["HeterogeneousCore/SonicTriton/test/tritonTest_cfg.py"],"time":1706298398},"152b2e0995d6d329cbce5753bb312dc85d567aca":{"files":["HeterogeneousCore/SonicTriton/test/README.md","HeterogeneousCore/SonicTriton/test/tritonTest_cfg.py","HeterogeneousCore/SonicTriton/test/unittest.sh"],"time":1706298399},"3ba5109cf05f8a9055bb0d46dfbe8dfe93fd083e":{"files":["HeterogeneousCore/SonicTriton/src/TritonService.cc"],"time":1706298396},"4b4b16f1f49eeb7240ca2412fc794aee1656c7a9":{"files":["HeterogeneousCore/SonicTriton/scripts/cmsTriton"],"time":1706299537},"6abef2c97bf0176177fda617c995ca67afeb808a":{"files":["HeterogeneousCore/SonicTriton/scripts/cmsTriton"],"time":1706298391},"6efa3a06f2ef8cb6b01bb89dd2e959166bf7dcad":{"files":["HeterogeneousCore/SonicTriton/scripts/cmsTriton"],"time":1706298393},"81c325914c2c4d89d0eee7683eb7a69822ea2b25":{"files":["HeterogeneousCore/SonicTriton/README.md","HeterogeneousCore/SonicTriton/src/TritonClient.cc","HeterogeneousCore/SonicTriton/test/tritonTest_cfg.py"],"time":1706298397},"86001328a18186c96e94f1406d7b4c7638345f97":{"files":["HeterogeneousCore/SonicTriton/README.md","HeterogeneousCore/SonicTriton/scripts/cmsTriton"],"time":1706298393},"87e2dd482c6edf1ca145c0d72eab6d05810a573c":{"files":["HeterogeneousCore/SonicTriton/interface/TritonClient.h","HeterogeneousCore/SonicTriton/interface/triton_utils.h","HeterogeneousCore/SonicTriton/src/TritonClient.cc","HeterogeneousCore/SonicTriton/src/TritonMemResource.cc","HeterogeneousCore/SonicTriton/src/TritonService.cc"],"time":1706298397},"9ada2b0899ae0b82de448aa0eae9e0fce94731bf":{"files":["Configuration/PyReleaseValidation/scripts/runTheMatrix.py"],"time":1706298394},"d0d4882b9b7ca1e10494bb21a8bd149a1e792867":{"files":["Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py"],"time":1706298394},"d24a6f4e02c8d5ec3129b535603a2f188f5c0368":{"files":["HeterogeneousCore/SonicTriton/interface/TritonException.h","HeterogeneousCore/SonicTriton/interface/TritonService.h","HeterogeneousCore/SonicTriton/interface/triton_utils.h","HeterogeneousCore/SonicTriton/src/TritonException.cc","HeterogeneousCore/SonicTriton/src/TritonService.cc"],"time":1706298395},"eab88918848d2a6ae1d085b25003da0679d3acd5":{"files":["HeterogeneousCore/SonicTriton/scripts/cmsTriton"],"time":1706298392},"emoji":{},"f6a5b7de52df737e6c54d3c9b9bfb25fee98790e":{"files":["HeterogeneousCore/SonicTriton/interface/TritonException.h","HeterogeneousCore/SonicTriton/src/TritonClient.cc","HeterogeneousCore/SonicTriton/src/TritonException.cc"],"time":1706298396}} -->

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43814/38622

- This PR adds an extra 124KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532, #43294, #43701, #43761, #43805

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43814/38622/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43814/38622/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43814/38624

- This PR adds an extra 120KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532, #43294, #43701, #43761, #43805

cmsbuild commented:
A new Pull Request was created by  @kpedro88 (Kevin Pedro) for master.

It involves the following packages:

- Configuration/PyReleaseValidation (**pdmv, upgrade**)
- HeterogeneousCore/SonicTriton (**heterogeneous**)


@miquork, @subirsarkar, @makortel, @AdrianoDee, @fwyzard, @cmsbuild, @sunilUIET, @srimanob can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @riga, @missirol, @rovere, @makortel, @fabiocos, @slomeo this is something you requested to watch as well.
@antoniovilela, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented on HeterogeneousCore/SonicTriton/src/TritonService.cc:
```diff
@@ -297,20 +300,39 @@ void TritonService::preBeginJob(edm::PathsAndConsumesOfModulesBase const&, edm::
+    if (rv != 0) {
+      cms::Exception stopException("FallbackFailed");
+      stopException << "TritonService: Stopping the fallback server failed with exit code " << rv;
+      //avoid throwing if the stack is already unwinding
+      if (callFails_ > 0)
+        edm::LogWarning(stopException.category()) << stopException.explainSelf();
+      else
+        throw stopException;
+    }
```
From the framework point of view you are technically allowed to throw an exception also in the case where the framework is shutting down because of an exception. In that case the framework has caught the exception and stored it, and strictly speaking the C++ runtime is not in a "stack unwinding" state when the framework calls `postEndJob()`. On the other hand, framework ignores all further exceptions from `endJob()`, so (at least for now) it probably makes sense to keep the printout for that case.

Thinking further about production use, we generally would not want errors in `endJob()` to lead to job failures, because the event processing itself finished. On the other hand, failing to stop the fallback server sounds a bit tricky situation, and the desired behavior might need input from WM. In the IBs I'd imagine we'd want to see these failures, so for time being I think the existing exception here is fine.

makortel replied:
> In that case the framework has caught the exception and stored it, and strictly speaking the C++ runtime is not in a "stack unwinding" state when the framework calls `postEndJob()`

I'm wrong here, in this case the `postEndJob()` is called as part of the stack unwinding of the first exception. But it is called within a specific `try`-`catch` block (actually nested twice), so there won't be two uncaught exceptions in flight (that would lead to immediate `std::terminate()`).

kpedro88 replied:
@makortel so is the conclusion that converting from exception to warning is not needed here?

kpedro88 committed:
fix handling of symbolic links

kpedro88 commented:
@makortel @Dr15Jones there is another issue about multiple exceptions, #38260. I haven't tried to address that here because I'm not sure what the recommendation is. Any advice?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43814/38647

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532, #43701, #43761, #43805

cmsbuild commented:
Pull request #43814 was updated. @sunilUIET, @srimanob, @subirsarkar, @fwyzard, @AdrianoDee, @makortel, @miquork, @cmsbuild can you please check and sign again.

Dr15Jones commented:
> there is another issue about multiple exceptions, https://github.com/cms-sw/cmssw/issues/38260. I haven't tried to address that here because I'm not sure what the recommendation is. Any advice?

Have the destructor for TritonCpuShmResource catch all exceptions.

makortel commented on HeterogeneousCore/SonicTriton/src/TritonService.cc:
```diff
@@ -297,20 +300,39 @@ void TritonService::preBeginJob(edm::PathsAndConsumesOfModulesBase const&, edm::
-    throw cms::Exception("FallbackFailed")
-        << "TritonService: Stopping the fallback server failed with exit code " << rv;
+    if (rv != 0) {
+      cms::Exception stopException("FallbackFailed");
+      stopException << "TritonService: Stopping the fallback server failed with exit code " << rv;
```
I'd suggest to construct a string here that is used in the `LogWarning` and in the exception throwing, rather than using `cms::Exception` as a "string stream" for the `LogWarning`.

kpedro88 replied:
Done

makortel commented on HeterogeneousCore/SonicTriton/interface/TritonService.h:
```diff
@@ -132,6 +134,7 @@ class TritonService {
   unsigned currentModuleId_;
   bool allowAddModel_;
   bool startedFallback_;
+  CMS_SA_ALLOW mutable std::atomic<int> callFails_;
```
Wouldn't `bool` along
```suggestion
  CMS_SA_ALLOW mutable std::atomic<bool> anyCallFailed_{false};
```
be sufficient? (only set to `true` if any of the inference calls failed)

kpedro88 replied:
Whenever a call fails because of a potential local fallback server issue, a `TritonException` is created with the instruction to notify the `TritonService` that the call failed. But calls can be retried, and when a call is retried, it has to undo that notification. Only allowing `true` and `false` as possibilities means that the notification from a completely-failed call could get overwritten by a retrying call (depending on how these notifications are interleaved when running on multiple threads). Maybe there is a smarter way to do this, but an atomic counter seemed appropriate here.

makortel replied:
Ok. I forgot earlier to mention the `CMS_SA_ALLOW` is not needed for `std::atomic` (because static analyzer knows `std::atomic` is fully thread safe)
```suggestion
  mutable std::atomic<int> callFails_;
```

kpedro88 committed:
common message elements as string(stream)s

kpedro88 committed:
avoid throwing multiple exceptions in TritonMemResource destructors

kpedro88 commented:
@Dr15Jones I think [ccb9dee](https://github.com/cms-sw/cmssw/pull/43814/commits/ccb9dee9010c6c86e2980bb489d0b6612537bd23) should do the trick.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43814/38654

- This PR adds an extra 32KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532, #43701, #43761, #43805

cmsbuild commented:
Pull request #43814 was updated. @fwyzard, @srimanob, @AdrianoDee, @miquork, @cmsbuild, @subirsarkar, @makortel, @sunilUIET can you please check and sign again.

kpedro88 commented:
test parameters:
workflows = 10805.31,11634.9001,24834.9001
relvals_opt = --what cleanedupgrade,standard,highstats,pileup,generator,extendedgen,production,identity,ged,machine,premix,nano,gpu,2017,2026

kpedro88 commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4df92d/37166/summary.html
**COMMIT**: ccb9dee9010c6c86e2980bb489d0b6612537bd23
**CMSSW**: CMSSW_14_0_X_2024-02-01-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/43814/37166/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestHeterogeneousCoreSonicTritonVersionCheck had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [removed 106 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-02-01-1100+4df92d/60843/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 62 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-02-01-1100+4df92d/60843/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3248620
 - DQMHistoTests: [Total failures: 1192](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-02-01-1100+4df92d/60843/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3247406
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-02-01-1100+4df92d/60843/triggerResults): no differences found

kpedro88 committed:
another fix for symbolic link handling

kpedro88 commented:
Fixed unexpected unit test failure. The new unit test actually fully works now, and correctly reports that I missed a cms-data directory that needed checksums added: now addressed in https://github.com/cms-data/RecoTauTag-TrainingFiles/pull/13.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43814/38667

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532, #43701, #43761, #43805
  - File  Configuration/PyReleaseValidation/scripts/runTheMatrix.py  modified in PR(s): #43824

cmsbuild commented:
Pull request #43814 was updated. @sunilUIET, @fwyzard, @AdrianoDee, @miquork, @makortel, @subirsarkar, @srimanob, @cmsbuild can you please check and sign again.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4df92d/37206/summary.html
**COMMIT**: 1c86d910aa7b620332f32b1f24a481478c14bd67
**CMSSW**: CMSSW_14_0_X_2024-02-04-0000/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43814/37206/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 98 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-02-04-0000+4df92d/60881/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 68 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-02-04-0000+4df92d/60881/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3248614
 - DQMHistoTests: [Total failures: 1189](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-02-04-0000+4df92d/60881/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3247403
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_0_X_2024-02-04-0000+4df92d/60881/triggerResults): no differences found

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_14_1_X. Please open a backport if it should also go in to CMSSW_14_0_X.

makortel reviewed (commented):
These are my last comments.

kpedro88 committed:
code review

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-43814/38740

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #33532, #43701, #43761, #43805
  - File  Configuration/PyReleaseValidation/scripts/runTheMatrix.py  modified in PR(s): #43824

cmsbuild commented:
Pull request #43814 was updated. @srimanob, @subirsarkar, @makortel, @cmsbuild, @sunilUIET, @fwyzard, @miquork, @AdrianoDee can you please check and sign again.

kpedro88 commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4df92d/37258/summary.html
**COMMIT**: ec8dda148e52b0e963f525a8638ea5c238f21602
**CMSSW**: CMSSW_14_1_X_2024-02-06-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/43814/37258/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 6 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-06-1100+4df92d/60923/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 57 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-06-1100+4df92d/60923/validateJR.html)
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3248614
 - DQMHistoTests: [Total failures: 1189](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-06-1100+4df92d/60923/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3247403
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 200 log files, 161 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_1_X_2024-02-06-1100+4df92d/60923/triggerResults): no differences found

makortel commented:
The usual
* https://github.com/cms-sw/cmssw/issues/39803
* https://github.com/cms-sw/cmssw/issues/43293
* https://github.com/cms-sw/cmssw/issues/43790

makortel commented:
+heterogeneous

srimanob commented:
+Upgrade

kpedro88 commented:
@cms-sw/pdmv-l2 please sign (from your side, just expanding a special workflow to another fragment)

AdrianoDee commented:
+pdmv

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @antoniovilela, @sextonkennedy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

antoniovilela commented:
+1
